node/src/chain/fsm.rs

```diff
@@ -209,14 +178,88 @@ impl<N: Network, DB: database::DB, VM: vm::VMExecution> SimpleFSM<N, DB, VM> {
                     }
                 }
             }
-        }
 
+            // Try to detect a stalled chain
+            // Generally speaking, if a node is receiving future blocks from the
+            // network but it cannot accept a new block for long time, then
+            // it might be a sign of a getting stalled on non-main branch.
+
+            let res = self.stalled_sm.on_block_received(blk).await.clone();
+
+            match res {
+                stall_chain_fsm::State::StalledOnFork(
+                    local_hash_at_fork,
+                    remote_blk,
+                ) => {
+                    info!(
+                        event = "stalled on fork",
+                        local_hash = to_str(&local_hash_at_fork),
+                        remote_hash = to_str(&remote_blk.header().hash),
+                        remote_height = remote_blk.header().height,
+                    );
+                    let mut acc = self.acc.write().await;
+
+                    let prev_local_state_root =
+                        acc.db.read().await.view(|t| {
+                            let local_blk = t
+                                .fetch_block_header(&local_hash_at_fork)?
+                                .expect("local hash should exist");
+
+                            let prev_blk = t
+                                .fetch_block_header(&local_blk.prev_block_hash)?
+                                .expect("prev block hash should exist");
+
+                            anyhow::Ok(prev_blk.state_hash)
+                        })?;
+
+                    match acc
+                        .try_revert(RevertTarget::Commit(prev_local_state_root))
+                        .await
+                    {
+                        Ok(_) => {
+                            counter!("dusk_revert_count").increment(1);
+                            info!(event = "reverted to last finalized");
+
+                            info!(
+                                event = "recovery block",
+                                height = remote_blk.header().height,
+                                hash = to_str(&remote_blk.header().hash),
+                            );
+
+                            acc.try_accept_block(&remote_blk, true).await?;
+
+                            // Black list the block hash to avoid accepting it
+                            // again due to fallback execution
+                            self.blacklisted_blocks
+                                .write()
+                                .await
+                                .insert(local_hash_at_fork);
+
```

Comment:
to speedup, we might request the following blocks from the sender, instead of waiting for the next quorum. 
if not in this pr, maybe let's remember it for a future improvement

Reply:
if we just request missing blocks will not be an improvement. We need to switch to `pre-sync` and then to `out-of-sync` state.

Reply:
yeah, that's what I meant. but let's do it in a future pr